server/src/main/java/org/elasticsearch/index/translog/BaseTranslogReader.java

```diff
@@ -107,11 +108,19 @@ protected final BufferedChecksumStreamInput checksummedStream(ByteBuffer reusabl
         }
         buffer.clear();
         buffer.limit(opSize);
-        readBytes(buffer, position);
+        readBytesDetectingTruncation(buffer, position);
         buffer.flip();
         return new BufferedChecksumStreamInput(new ByteBufferStreamInput(buffer), path.toString(), reuse);
     }
 
+    private void readBytesDetectingTruncation(ByteBuffer buffer, long position) throws IOException {
```

Comment:
I slightly prefer not to introduce this new method. I am unsure that we will always use this method instead of "readBytes". Both methods have the same signature. How about catching EOFException in TranslogSnapshot#readOperation as you did for the translog header?

Reply:
I know what you mean, but wrapping this exception up in `TranslogSnapshot` means the behaviour there would differ from the other subclasses, `TranslogReader` and `TranslogWriter`. I could adjust all the `readBytes` methods to throw `TranslogCorruptedException` instead of `EOFException` perhaps?

Reply:
We can make `TranslogSnapshot#readOperation` final. I am fine with making `readBytes` throw TranslogCorruptedException instead of EOFException.

Reply:
Ok, I think you're saying that we only need to handle `EOFException` specially in `TranslogSnapshot`? I think that's true for the `elasticsearch-shard` tool, and have adjusted it in b51abde. I'm a little uncomfortable with the difference in behaviour between `TranslogSnapshot` and the other subclasses, `TranslogReader` and `TranslogWriter`, but I look to your judgement here.